src/CodeStyle/Core/CodeFixes/AbstractFormattingAnalyzerImpl.cs

```diff
@@ -0,0 +1,94 @@
+        {
+            var tree = context.Tree;
+            var cancellationToken = context.CancellationToken;
+
+            var options = workspace.Options;
+            if (File.Exists(tree.FilePath))
+            {
+                var codingConventionContext = codingConventionsManager.GetConventionContextAsync(tree.FilePath, cancellationToken).GetAwaiter().GetResult();
+                options = ApplyFormattingOptions(options, codingConventionContext);
+            }
+
+            var oldText = tree.GetText(cancellationToken);
+            var formattingChanges = Formatter.GetFormattedTextChanges(tree.GetRoot(cancellationToken), workspace, options, cancellationToken);
+
+            // formattingChanges could include changes that impact a larger section of the original document than
+            // necessary. Before reporting diagnostics, process the changes to minimize the span of individual
+            // diagnostics.
+            foreach (var formattingChange in formattingChanges)
+            {
+                var change = formattingChange;
+                if (change.NewText.Length > 0 && !change.Span.IsEmpty)
+                {
+                    // Handle cases where the change is a substring removal from the beginning. In these cases, we want
+                    // the diagnostic span to cover the unwanted leading characters (which should be removed), and
+                    // nothing more.
+                    var offset = change.Span.Length - change.NewText.Length;
+                    if (offset >= 0)
+                    {
+                        if (oldText.GetSubText(new TextSpan(change.Span.Start + offset, change.NewText.Length)).ContentEquals(SourceText.From(change.NewText)))
+                        {
+                            change = new TextChange(new TextSpan(change.Span.Start, offset), "");
+                        }
+                        else
+                        {
+                            // Handle cases where the change is a substring removal from the end. In these cases, we want
+                            // the diagnostic span to cover the unwanted trailing characters (which should be removed), and
+                            // nothing more.
+                            if (oldText.GetSubText(new TextSpan(change.Span.Start, change.NewText.Length)).ContentEquals(SourceText.From(change.NewText)))
+                            {
+                                change = new TextChange(new TextSpan(change.Span.Start + change.NewText.Length, offset), "");
+                            }
+                        }
+                    }
+                }
+
+                if (change.NewText.Length == 0 && change.Span.IsEmpty)
+                {
+                    // No actual change
+                    throw new InvalidOperationException("This program location is thought to be unreachable.");
```

Comment:
>throw new InvalidOperationException("This program location is thought to be unreachable.") [](start = 20, length = 90)

[ExceptionUtilities.Unreachable](http://source.roslyn.io/#Microsoft.CodeAnalysis/InternalUtilities/ExceptionUtilities.cs,24)?

Reply:
`ExceptionUtilities` does not exist in the code style layer (this layer has equal access to third-party analyzers, i.e. no IVT at all). I could redefine the extension in this layer or keep this as-is.

Reply:
Ah, then it is fine.

---
In reply to: [222051245](https://github.com/dotnet/roslyn/pull/28873#discussion_r222051245) [](ancestors = 222051245)